Copter: Zero throttle and relax - factor out common code from flight run functions

peterbarker opened the pull request:
From gitter: I'm factoring some code out of each of the control mode's run function. Can anybody tell me, of the dozen-or-so places that set a heli's calm-down function in the case we shouldn't really be in a controller, https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/control_land.cpp#L120 sets the roll, pitch and yaw to non-zero values? It seems to me that might be a bad thing.

I expect all commits to be squished together.

The numbers in brackets are the cksum(1)s of the blocks that have been replaced with a function call.

peterbarker commented:
I'm open to better names for the pulled-out-function.  "act_as_if_landed()"?

OXINARF commented on ArduCopter/flight_mode.cpp:
```diff
@@ -427,3 +427,15 @@ void Copter::print_flight_mode(AP_HAL::BetterStream *port, uint8_t mode)
+#else
+        motors.set_desired_spool_state(AP_Motors::DESIRED_SPIN_WHEN_ARMED);
+        // multicopters do not stabilize roll/pitch/yaw when disarmed
+        attitude_control.set_throttle_out_unstabilized(0,true,g.throttle_filt);
```
Nitpicking (but since we are moving code around):

- put space between the parameters
- change the multicopter comment to be in the same line as the else
- remove the *call attitude controller* comment, it's completely redundant

OXINARF commented:
LGTM, including the name (the alternative looks worse to me). I think that the heli doubt should be answered before merging this, otherwise it will be forgotten.

peterbarker commented:
@OXINARF The reason I was suggesting we might need a better name is that there are other bits of random cleanups that *some* of these places do - resetting I terms and wp navigation, for example.  It is possible that *all* of these places should be doing that, and that it should go into this function.  In that case it would start doing more than what the current name implies :-)

OXINARF commented:
@peterbarker Sure, if more stuff is added to this function we can change the name then. I think that for now the name is appropriate.

R-Lefebvre commented:
But otherwise, I think this PR LGTM.

OXINARF commented:
> I think we should consider including set_yaw_target_to_current_heading inside the zero_throttle_and_relax_ac function.
>
> It would mean this isn't a simple refactoring, we are slightly changing the functionality. But it might be the correct thing to do.

It would need Randy to validate that, delaying this further. Small steps are simpler to get in.

But I agree with you that Land without GPS is more similar to AltHold, except you don't control the throttle. In that sense I think the multirotor part may be wrong, as with AltHold even when in landed state it uses pilot input (https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/control_althold.cpp#L130). I'll ask Randy to look at this.

amilcarlucas commented:
ping

khancyr commented:
ping up

peterbarker commented:
@amilcarlucas @khancyr I've pushed up another patch to this.  It changes behaviour in land_run_nogps - for the better in my opinion.

OXINARF commented:
@peterbarker: @R-Lefebvre explained above why the Land NoGPS option should still allow the input to be taken into account in Heli. I even pointed out that, with multi, either AltHold mode is wrong or it looks like Land NoGPS is wrong - I think they should do the same thing when disarmed/landed.

This really needs Heli and Copter control experts looking at this.

peterbarker commented:
@OXINARF so are you suggesting we drop off the last patch?

If we can just get the factoring parts in that's a start.

OXINARF commented:
> @OXINARF so are you suggesting we drop off the last patch?
>
> If we can just get the factoring parts in that's a start.

Yes. I agree with getting the factoring in, that looks harmless. I also believe that doing this factoring will help for the other parts - I'm sure there are several parts of code that aren't doing the same when they should.

peterbarker commented:
@OXINARF I've popped off the landing patch and force-pushed this up.  Should just be refactoring now.

OXINARF commented on ArduCopter/mode.cpp:
```diff
@@ -262,3 +261,16 @@ bool Copter::Mode::takeoff_triggered(const float target_climb_rate) const
+#if FRAME_CONFIG == HELI_FRAME  // Helicopters always stabilize roll/pitch/yaw
+        // call attitude controller
+        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(0, 0, 0, get_smoothing_gain());
+        attitude_control->set_throttle_out(0,false,g.throttle_filt);
```
Nitpicking (but I'm a bit tired of PRs fixing it later): use `0.0f` and add spaces between function arguments. This applies to the two above and another below.

I know you copied this from previous code but since we are moving it, we might as well fix it already 🙂

peterbarker replied:
You forgot to tell me to fix the indentation :-P

peterbarker commented:
@OXINARF Rebased this branched and corrected the style of `zero_throttle_and_relax`.

I'm afraid I also went through and replaced some more code with `zero_throttle_and_relax`, however - patch on top.

OXINARF commented on ArduCopter/mode_drift.cpp:
```diff
@@ -48,8 +48,7 @@ void Copter::ModeDrift::run()
     if (!motors->armed() || !motors->get_interlock() || (ap.land_complete && ap.throttle_zero)) {
-        motors->set_desired_spool_state(AP_Motors::DESIRED_SPIN_WHEN_ARMED);
-        attitude_control->set_throttle_out_unstabilized(0,true,g.throttle_filt);
+        zero_throttle_and_relax_ac();
```
Here I don't know, this introduces new behaviour for helis.

OXINARF commented on ArduCopter/mode_land.cpp:
```diff
@@ -108,16 +108,7 @@ void Copter::ModeLand::nogps_run()
-        attitude_control->set_throttle_out_unstabilized(0,true,g.throttle_filt);
-#endif
-
+        zero_throttle_and_relax_ac();
```
Humm, so you brought this one back? As I wrote earlier, this change looks wrong to me - it's actually multi doing the wrong thing here.

OXINARF commented on ArduCopter/mode_poshold.cpp:
```diff
@@ -140,9 +140,8 @@ void Copter::ModePosHold::run()
-        motors->set_desired_spool_state(AP_Motors::DESIRED_SPIN_WHEN_ARMED);
         wp_nav->init_loiter_target();
-        attitude_control->set_throttle_out_unstabilized(0,true,g.throttle_filt);
+        zero_throttle_and_relax_ac();
```
Again, changing behaviour for helis. Is it wrong or right? I have no idea.

peterbarker commented:
I've reverted the last patch on this and put those changes into a separate PR.

OXINARF commented:
Merged, thanks!

OXINARF commented:
Merged, thanks!
